File: libs/shape/src/test/kotlin/mono/shape/shape/LineTest.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.shape

import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.Point
import mono.shape.serialization.SerializableLine
import mono.shape.shape.line.LineHelper
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * A test for [Line]
 */
class LineTest {
    @Test
    fun testSerialization_init() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        assertEquals(startPoint, serializableLine.startPoint)
        assertEquals(endPoint, serializableLine.endPoint)
        assertEquals(line.reducedJoinPoints, LineHelper.reduce(serializableLine.jointPoints))
        assertEquals(line.extra.toSerializableExtra(), serializableLine.extra)
        assertFalse(serializableLine.wasMovingEdge)
    }

    @Test
    fun testSerialization_moveAnchorPoints() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)

        val newStartPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 4, 5)
        val newEndPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 7, 8)
        line.moveAnchorPoint(
            Line.AnchorPointUpdate(Line.Anchor.START, newStartPoint),
            isReduceRequired = true
        )
        line.moveAnchorPoint(
            Line.AnchorPointUpdate(Line.Anchor.END, newEndPoint),
            isReduceRequired = true
        )

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        assertEquals(newStartPoint, serializableLine.startPoint)
        assertEquals(newEndPoint, serializableLine.endPoint)
        assertEquals(line.reducedJoinPoints, LineHelper.reduce(serializableLine.jointPoints))
        assertEquals(line.extra.toSerializableExtra(), serializableLine.extra)
        assertFalse(serializableLine.wasMovingEdge)
    }

    @Test
    fun testSerialization_moveEdge() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)
        line.moveEdge(line.edges.first().id, Point(10, 10), isReduceRequired = true)

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        assertEquals(startPoint, serializableLine.startPoint)
        assertEquals(endPoint, serializableLine.endPoint)
        assertEquals(line.reducedJoinPoints, LineHelper.reduce(serializableLine.jointPoints))
        assertEquals(line.extra.toSerializableExtra(), serializableLine.extra)
        assertTrue(serializableLine.wasMovingEdge)
    }

    @Test
    fun testSerialization_restoreInit() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        val line2 = Line(serializableLine, parentId = PARENT_ID)
        assertEquals(line.getDirection(Line.Anchor.START), line2.getDirection(Line.Anchor.START))
        assertEquals(line.getDirection(Line.Anchor.END), line2.getDirection(Line.Anchor.END))
        assertEquals(line.reducedJoinPoints, line2.reducedJoinPoints)
        assertEquals(line.extra, line2.extra)
        assertEquals(line.wasMovingEdge(), line2.wasMovingEdge())
    }

    @Test
    fun testSerialization_restoreAfterMovingAnchorPoints() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)

        val newStartPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 4, 5)
        val newEndPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 7, 8)
        line.moveAnchorPoint(
            Line.AnchorPointUpdate(Line.Anchor.START, newStartPoint),
            isReduceRequired = true
        )
        line.moveAnchorPoint(
            Line.AnchorPointUpdate(Line.Anchor.END, newEndPoint),
            isReduceRequired = true
        )

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        val line2 = Line(serializableLine, parentId = PARENT_ID)
        assertEquals(line.getDirection(Line.Anchor.START), line2.getDirection(Line.Anchor.START))
        assertEquals(line.getDirection(Line.Anchor.END), line2.getDirection(Line.Anchor.END))
        assertEquals(line.reducedJoinPoints, line2.reducedJoinPoints)
        assertEquals(line.extra, line2.extra)
        assertEquals(line.wasMovingEdge(), line2.wasMovingEdge())
    }

    @Test
    fun testSerialization_restoreAfterMovingEdge() {
        val startPoint = DirectedPoint(DirectedPoint.Direction.VERTICAL, 1, 2)
        val endPoint = DirectedPoint(DirectedPoint.Direction.HORIZONTAL, 3, 4)
        val line = Line(startPoint, endPoint, parentId = PARENT_ID)
        line.moveEdge(line.edges.first().id, Point(10, 10), isReduceRequired = true)

        val serializableLine = line.toSerializableShape(true) as SerializableLine
        val line2 = Line(serializableLine, PARENT_ID)
        assertEquals(line.getDirection(Line.Anchor.START), line2.getDirection(Line.Anchor.START))
        assertEquals(line.getDirection(Line.Anchor.END), line2.getDirection(Line.Anchor.END))
        assertEquals(line.reducedJoinPoints, line2.reducedJoinPoints)
        assertEquals(line.extra, line2.extra)
        assertEquals(line.wasMovingEdge(), line2.wasMovingEdge())
    }

    companion object {
        private const val PARENT_ID = "1"
    }
}

File: monosketch-svelte/src/lib/mono/shape/shape/line.test.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import type { SerializableLine } from "$mono/shape/serialization/shapes";
import { describe, it, expect } from 'vitest';
import { DirectedPoint, Direction, Point } from '$libs/graphics-geo/point';
import { Line } from '$mono/shape/shape/line';
import { LineAnchor, LineHelper } from '$mono/shape/shape/linehelper';

describe('Line', () => {
    const PARENT_ID = "1";

    it('testSerialization_init', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        expect(serializableLine.startPoint).toEqual(startPoint);
        expect(serializableLine.endPoint).toEqual(endPoint);
        expect(LineHelper.reduce(serializableLine.jointPoints)).toEqual(line.reducedJoinPoints);
        expect(serializableLine.extra).toEqual(line.extra.toSerializableExtra());
        expect(serializableLine.wasMovingEdge).toBe(false);
    });

    it('testSerialization_moveAnchorPoints', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);

        const newStartPoint = DirectedPoint.of(Direction.HORIZONTAL, 4, 5);
        const newEndPoint = DirectedPoint.of(Direction.VERTICAL, 7, 8);
        line.moveAnchorPoint(
            { anchor: LineAnchor.START, point: newStartPoint },
            true,
            false,
        );
        line.moveAnchorPoint(
            { anchor: LineAnchor.END, point: newEndPoint },
            true,
            false,
        );

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        expect(serializableLine.startPoint).toEqual(newStartPoint);
        expect(serializableLine.endPoint).toEqual(newEndPoint);
        expect(LineHelper.reduce(serializableLine.jointPoints)).toEqual(line.reducedJoinPoints);
        expect(serializableLine.extra).toEqual(line.extra.toSerializableExtra());
        expect(serializableLine.wasMovingEdge).toBe(false);
    });

    it('testSerialization_moveEdge', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);
        line.moveEdge(line.edges[0].id, new Point(10, 10), true);

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        expect(serializableLine.startPoint).toEqual(startPoint);
        expect(serializableLine.endPoint).toEqual(endPoint);
        expect(LineHelper.reduce(serializableLine.jointPoints)).toEqual(line.reducedJoinPoints);
        expect(serializableLine.extra).toEqual(line.extra.toSerializableExtra());
        expect(serializableLine.wasMovingEdge).toBe(true);
    });

    it('testSerialization_restoreInit', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        const line2 = Line.fromSerializable(serializableLine, PARENT_ID);
        expect(line2.getDirection(LineAnchor.START)).toEqual(line.getDirection(LineAnchor.START));
        expect(line2.getDirection(LineAnchor.END)).toEqual(line.getDirection(LineAnchor.END));
        expect(line2.reducedJoinPoints).toEqual(line.reducedJoinPoints);
        expect(line2.extra).toEqual(line.extra);
        expect(line2.wasMovingEdge()).toEqual(line.wasMovingEdge());
    });

    it('testSerialization_restoreAfterMovingAnchorPoints', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);

        const newStartPoint = DirectedPoint.of(Direction.HORIZONTAL, 4, 5);
        const newEndPoint = DirectedPoint.of(Direction.VERTICAL, 7, 8);
        line.moveAnchorPoint(
            { anchor: LineAnchor.START, point: newStartPoint },
            true,
            false,
        );
        line.moveAnchorPoint(
            { anchor: LineAnchor.END, point: newEndPoint },
            true,
            false,
        );

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        const line2 = Line.fromSerializable(serializableLine, PARENT_ID);
        expect(line2.getDirection(LineAnchor.START)).toEqual(line.getDirection(LineAnchor.START));
        expect(line2.getDirection(LineAnchor.END)).toEqual(line.getDirection(LineAnchor.END));
        expect(line2.reducedJoinPoints).toEqual(line.reducedJoinPoints);
        expect(line2.extra).toEqual(line.extra);
        expect(line2.wasMovingEdge()).toEqual(line.wasMovingEdge());
    });

    it('testSerialization_restoreAfterMovingEdge', () => {
        const startPoint = DirectedPoint.of(Direction.VERTICAL, 1, 2);
        const endPoint = DirectedPoint.of(Direction.HORIZONTAL, 3, 4);
        const line = new Line(startPoint, endPoint, undefined, PARENT_ID);
        line.moveEdge(line.edges[0].id, new Point(10, 10), true);

        const serializableLine = line.toSerializableShape(true) as SerializableLine;
        const line2 = Line.fromSerializable(serializableLine, PARENT_ID);
        expect(line2.getDirection(LineAnchor.START)).toEqual(line.getDirection(LineAnchor.START));
        expect(line2.getDirection(LineAnchor.END)).toEqual(line.getDirection(LineAnchor.END));
        expect(line2.reducedJoinPoints).toEqual(line.reducedJoinPoints);
        expect(line2.extra).toEqual(line.extra);
        expect(line2.wasMovingEdge()).toEqual(line.wasMovingEdge());
    });
});